### PR TITLE
NETOBSERV-1545: Expose a counter for BPF hashmap update packets drop

### PR DESCRIPTION
**Pull request by msherif1234**
## Description

cilium doesn't seem to have a way to read globals from ebpf program I will keep this PR as draft till we have a way to read global from userspace

https://cilium.slack.com/archives/C027KBX679U/p1711370379468299

so perCPU array map will be used to hold the global counter and userspace will read, aggregate and update the metrics

## Dependencies

n/a

## Checklist

If you are not familiar with our processes or don't know what to answer in the list below, let us know in a comment: the maintainers will take care of that.

* [ ] Will this change affect NetObserv / Network Observability operator? If not, you can ignore the rest of this checklist.
* [ ] Is this PR backed with a JIRA ticket? If so, make sure it is written as a title prefix _(in general, PRs affecting the NetObserv/Network Observability product should be backed with a JIRA ticket - especially if they bring user facing changes)._
* [ ] Does this PR require product documentation?
  * [ ] If so, make sure the JIRA epic is labelled with "documentation" and provides a description relevant for doc writers, such as use cases or scenarios. Any required step to activate or configure the feature should be documented there, such as new CRD knobs.
* [ ] Does this PR require a product release notes entry?
  * [ ] If so, fill in "Release Note Text" in the JIRA.
* [ ] Is there anything else the QE team should know before testing? E.g: configuration changes, environment setup, etc.
  * [ ] If so, make sure it is described in the JIRA ticket.
* QE requirements (check 1 from the list):
  * [ ] Standard QE validation, with pre-merge tests unless stated otherwise.
  * [ ] Regression tests only (e.g. refactoring with no user-facing change).
  * [ ] No QE (e.g. trivial change with high reviewer's confidence, or per agreement with the QE team).

### CODE REVIEW

**Comment by msherif1234** [comment]
image

File: bpf/flows.c [hunk 1]
```diff
@@ -89,13 +89,24 @@ static inline int flow_monitor(struct __sk_buff *skb, u8 direction) {
...
             // In this case, the flow is dropped, as submitting it to the ringbuffer would cause
             // a duplicated UNION of flows (two different flows with partial aggregation of the same packets),
             // which can't be deduplicated.
             // other possible values https://chromium.googlesource.com/chromiumos/docs/+/master/constants/errnos.md
-            bpf_printk("error updating flow %d\n", ret);
+            if (trace_messages) {
+                bpf_printk("error updating flow %d\n", ret);
+            }
+            // Update global counter for hashmap update errors
+            error_counter_p = bpf_map_lookup_elem(&global_counters, &key);
```
Comment (jotak) [inline bpf/flows.c:104]: to make sure if I understand that: you created `global_counters` has a map for generic purpose, ie. today it contains drop counters at key 0 but potentially later we may add more counters at different indexes, is this correct?
Perhaps if so you could define a constant DROP_COUNTER_KEY = 0 or something like that?

File: pkg/ebpf/tracer.go [hunk 2]
```diff
@@ -448,9 +451,25 @@ func (m *FlowFetcher) LookupAndDeleteMap(met *metrics.Metrics) map[BpfFlowId][]B
...
 	met.BufferSizeGauge.WithBufferName("hashmap-unique").Set(float64(len(flows)))
 
+	m.ReadGlobalCounter(met)
 	return flows
 }
 
+// ReadGlobalCounter reads the global counter and updates hashmap update error counter metrics
+func (m *FlowFetcher) ReadGlobalCounter(met *metrics.Metrics) {
+	var allCPUValue []uint32
+	key := uint32(0)
```
Comment (jotak) [inline pkg/ebpf/tracer.go:461]: here as well this key could be a constant like `dropCounterKey = 0`

File: bpf/flows.c (hunk 1 above)
Comment (msherif1234) [reply bpf/flows.c:104]: yes and I am already thinking of using this to count filtered flows :) at different index as well as ovs dbg I will add const here and the golang side Thanks

File: pkg/ebpf/tracer.go [hunk 3]
```diff
@@ -448,9 +451,25 @@ func (m *FlowFetcher) LookupAndDeleteMap(met *metrics.Metrics) map[BpfFlowId][]B
...
+
+	if err := m.objects.GlobalCounters.Lookup(key, &allCPUValue); err != nil {
+		log.WithError(err).Warnf("couldn't read global counter")
+		return
+	}
+	// aggregate all the counters
+	for _, counter := range allCPUValue {
+		met.Errors.WithErrorName("flow-fetcher", "CannotUpdateHashMapCounter").Add(float64(counter))
```
Comment (jotak) [inline pkg/ebpf/tracer.go:469]: There's already a metric for drops, could we use it instead?

```suggestion
		met.DroppedFlowsCounter.WithSourceAndReason("flow-fetcher", "CannotUpdateHashMapCounter").Add(float64(counter))
```

Comment (msherif1234) [reply pkg/ebpf/tracer.go:469]: SG will use it then

**Review by jotak — APPROVED** [review]
thanks @msherif1234 !

**Pull request merged**